### PR TITLE
Dynamic resolution - support for SetDesktopSize in client

### PR DESCRIPTION
**Pull request by finlob**
Adds support for the client to request the server to change resolution.

Should resolve issue #341

### CODE REVIEW

File: libvncclient/rfbproto.c [hunk 1]
```diff
@@ -1392,7 +1394,11 @@ SendFramebufferUpdateRequest(rfbClient* client, int x, int y, int w, int h, rfbB
   rfbFramebufferUpdateRequestMsg fur;
 
   if (!SupportsClient2Server(client, rfbFramebufferUpdateRequest)) return TRUE;
-  
+  rfbClientLog("FramebufferUpdateRequest: %d,%d,%d,%d\n", x,y,w,h);
```
Comment (bk138) [inline libvncclient/rfbproto.c:1397]: This probably means a lot of messages. Please remove.

File: libvncclient/rfbproto.c [hunk 2]
```diff
@@ -1585,6 +1591,52 @@ SendKeyEvent(rfbClient* client, uint32_t key, rfbBool down)
...
+
+rfbBool
+SendExtDesktopSize(rfbClient* client, uint16_t width, uint16_t height)
+{
+  rfbSetDesktopSizeMsg sdm;
+  if (client->screen.width == 0 && client->screen.height == 0 ) {
+    rfbClientLog("Screen not yet received from server - not sending dimensions %dx%d\n", width, height);
+    return TRUE;
+  }
+  if (client->screen.width != rfbClientSwap16IfLE(width) && client->screen.height != rfbClientSwap16IfLE(height)) {
```
Comment (bk138) [inline libvncclient/rfbproto.c:1620]: Not sure why rfbClientSwap16IfLE is needed here, as width and height are not going over the network. Can you please explain?

File: libvncclient/rfbproto.c [hunk 3]
```diff
@@ -1585,6 +1591,52 @@ SendKeyEvent(rfbClient* client, uint32_t key, rfbBool down)
...
+    rfbClientLog("Sending dimensions %dx%d\n", width, height);
+    sdm.type = rfbSetDesktopSize;
+    sdm.width = rfbClientSwap16IfLE(width);
+    sdm.height = rfbClientSwap16IfLE(height);
+    sdm.numberOfScreens = 1;
+    client->screen.width = rfbClientSwap16IfLE(width);
```
Comment (bk138) [inline libvncclient/rfbproto.c:1626]: Shouldn't these be in host byte order?

File: libvncclient/rfbproto.c [hunk 4]
```diff
@@ -1585,6 +1591,52 @@ SendKeyEvent(rfbClient* client, uint32_t key, rfbBool down)
...
+{
+  rfbSetDesktopSizeMsg sdm;
+  if (client->screen.width == 0 && client->screen.height == 0 ) {
+    rfbClientLog("Screen not yet received from server - not sending dimensions %dx%d\n", width, height);
+    return TRUE;
+  }
+  if (client->screen.width != rfbClientSwap16IfLE(width) && client->screen.height != rfbClientSwap16IfLE(height)) {
+    rfbClientLog("Sending dimensions %dx%d\n", width, height);
+    sdm.type = rfbSetDesktopSize;
+    sdm.width = rfbClientSwap16IfLE(width);
```
Comment (bk138) [inline libvncclient/rfbproto.c:1623]: OK here, as it get's send over the network.

File: libvncclient/rfbproto.c [hunk 5]
```diff
@@ -1585,6 +1591,52 @@ SendKeyEvent(rfbClient* client, uint32_t key, rfbBool down)
...
+    sdm.height = rfbClientSwap16IfLE(height);
+    sdm.numberOfScreens = 1;
+    client->screen.width = rfbClientSwap16IfLE(width);
+    client->screen.height = rfbClientSwap16IfLE(height);
+
+    if (!WriteToRFBServer(client, (char *)&sdm, sz_rfbSetDesktopSizeMsg))
+      return FALSE;
+    if (!WriteToRFBServer(client, (char *)&(client->screen), sz_rfbExtDesktopScreen))
```
Comment (bk138) [inline libvncclient/rfbproto.c:1631]: Maybe it's better to have a temporary screen struct on the stack and write the BO-swapped values in there.

File: libvncclient/rfbproto.c [hunk 6]
```diff
@@ -1585,6 +1591,52 @@ SendKeyEvent(rfbClient* client, uint32_t key, rfbBool down)
...
+      return FALSE;
+
+    client->requestedResize = FALSE;
+
+    SendFramebufferUpdateRequest(client, 0, 0, 10, 10, FALSE);
```
Comment (bk138) [inline libvncclient/rfbproto.c:1636]: The values here seem bogus.

File: libvncclient/rfbproto.c [hunk 7]
```diff
@@ -1585,6 +1591,52 @@ SendKeyEvent(rfbClient* client, uint32_t key, rfbBool down)
   return WriteToRFBServer(client, (char *)&ke, sz_rfbKeyEventMsg);
 }
 
+/*
+ * Resize client
+ */
+rfbBool
```
Comment (bk138) [inline libvncclient/rfbproto.c:1597]: Can this be made static?

File: libvncclient/rfbproto.c [hunk 8]
```diff
@@ -2257,6 +2336,8 @@ HandleRFBServerMessage(rfbClient* client)
       }
     }
   }
+  if (!SendIncrementalFramebufferUpdateRequest(client))
```
Comment (bk138) [inline libvncclient/rfbproto.c:2339]: Why did this one move down here?

File: libvncclient/vncviewer.c [hunk 9]
```diff
@@ -357,6 +357,9 @@ rfbClient* rfbGetClient(int bitsPerSample,int samplesPerPixel,
   client->saslSecret = NULL;
 #endif /* LIBVNCSERVER_HAVE_SASL */
 
+  client->requestedResize = FALSE;
```
Comment (bk138) [inline libvncclient/vncviewer.c:360]: Should not be needed as we calloc() the struct.

**Review by bk138 — CHANGES_REQUESTED** [review]
Thanks for the contribution. A very good start, some comments though!

**Comment by antenore** [comment]
@finlob are you still interested in completing this PR?

The best I can offer at the moment is some testing.

Thanks!!!

**Comment by marcofortina** [comment]
Very interesting! Anyone is going on with this PR? Or is there a new one coming soon?

File: libvncclient/rfbproto.c (hunk 6 above)
Comment (marcofortina) [reply libvncclient/rfbproto.c:1636]: Is this right?

`SendFramebufferUpdateRequest(client, 0, 0, width, height, FALSE);`

File: libvncclient/rfbproto.c (hunk 1 above)
Comment (marcofortina) [reply libvncclient/rfbproto.c:1397]: I agree with @bk138 this introduce a lot of log messages, thus it should be removed

File: libvncclient/rfbproto.c (hunk 2 above)
Comment (marcofortina) [reply libvncclient/rfbproto.c:1620]: Hello @bk138
I did some tests and I think using rfbClientSwap16IfLE here is correct

Debug variables values:
client->screen.width: 21764
client->screen.height: 16131
width: 1109
height: 831
rfbClientSwap16IfLE(width): 21764
rfbClientSwap16IfLE(height): 16131

File: libvncclient/rfbproto.c (hunk 7 above)
Comment (marcofortina) [reply libvncclient/rfbproto.c:1597]: Tested and works also making this static

File: libvncclient/rfbproto.c (hunk 3 above)
Comment (marcofortina) [reply libvncclient/rfbproto.c:1626]: I think it's correct. See my comment above with debugging values information.

File: libvncclient/rfbproto.c (hunk 8 above)
Comment (marcofortina) [reply libvncclient/rfbproto.c:2339]: I tested moving this up in its original place and it works. IDK why it was moved down

**Pull request closed without merging**